x-pack/plugins/security/server/audit_trail/audit_trail_service.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Subject, combineLatest, of } from 'rxjs';
+import { filter } from 'rxjs/operators';
+import { KibanaRequest, Logger } from 'src/core/server';
+
+import {
+  AuditEvent,
+  AuditEventDecorator,
+  AuditorFactory,
+  LoggingServiceSetup,
+  AuditTrailSetup,
+  HttpServiceSetup,
+  OnPreResponseInfo,
+} from 'src/core/server';
+import { AuditTrailClient } from './audit_trail_client';
+
+import { SecurityPluginSetup } from '../';
+import { SpacesPluginSetup } from '../../../spaces/server';
+import { SecurityLicense } from '../../common/licensing';
+import { ConfigType } from '../config';
+
+interface SetupParams {
+  license: Pick<SecurityLicense, 'features$'>;
+  config: ConfigType['audit'];
+  getCurrentUser: SecurityPluginSetup['authc']['getCurrentUser'];
+  getSpacesService(): Pick<SpacesPluginSetup['spacesService'], 'getSpaceId'> | undefined;
+  logging: Pick<LoggingServiceSetup, 'configure'>;
+  auditTrail: Pick<AuditTrailSetup, 'register'>;
+  http: Pick<HttpServiceSetup, 'registerOnPreResponse'>;
+}
+
+export class AuditTrailService {
+  private readonly event$ = new Subject<AuditEvent>();
+
+  constructor(private readonly logger: Logger) {}
+
+  public setup({
+    license,
+    config,
+    logging,
+    http,
+    auditTrail,
+    getCurrentUser,
+    getSpacesService,
+  }: SetupParams) {
+    combineLatest(this.event$.asObservable(), license.features$)
+      .pipe(filter(([, licenseFeatures]) => licenseFeatures.allowAuditLogging))
+      .subscribe(([{ message, ...other }]) => this.logger.debug(message, other));
+
+    const depsApi = {
+      getCurrentUser,
+      getSpaceId: (request: KibanaRequest) => getSpacesService?.()?.getSpaceId(request)!,
+    };
+    const auditorFactory: AuditorFactory = {
+      asScoped: (request) => {
+        return new AuditTrailClient(request, this.event$, depsApi);
+      },
+    };
+    auditTrail.register(auditorFactory);
+
+    // TODO: Probably not the right hook since response can still change
+    http.registerOnPreResponse((request, preResponseInfo, t) => {
+      auditorFactory.asScoped(request).add(httpRequestEvent, { request, preResponseInfo });
+      return t.next();
+    });
+
+    logging.configure(
+      of({
+        appenders: {
+          auditTrailAppender: config.appender ?? {
+            kind: 'console',
+            layout: {
+              kind: 'pattern',
+              highlight: true,
+            },
+          },
+        },
+        loggers: [
+          {
+            context: 'audit_trail',
+            level: config.enabled ? 'debug' : 'off',
+            appenders: ['auditTrailAppender'],
+          },
+        ],
```

Comment:
This default appender/logger only activates if `verbose` logging is enabled (or if you are using the new undocumented logging config, if you have the `debug` log level enabled).

Adding onto what @legrego mentioned in https://github.com/elastic/kibana/pull/74640#discussion_r468098428, I think we need to consider what defaults we want to have here. We probably don't want the new audit logs to go to the console by default, as that's where the legacy audit logs are going.

Reply:
> or if you are using the new undocumented logging config

Let me know if we can improve logging config docs https://github.com/elastic/kibana/blob/master/src/core/server/logging/README.md

Reply:
@restrry sorry, I meant "undocumented" as in it's not on our docs website, but that was disingenuous! Thanks for linking the very well-written README here 😁

Reply:
> We probably don't want the new audit logs to go to the console by default, as that's where the legacy audit logs are going.

I thought we were going to enable at most one of the audit loggers, based on config settings. If that's still the plan, then writing to stdout by default wouldn't necessarily be a bad thing.


Taking inspiration from [what Elasticsearch does](https://www.elastic.co/guide/en/elasticsearch/reference/current/enable-audit-logging.html#enable-audit-logging), we could consider writing to `kibana_audit.json`, but that is not consistent with Kibana's default logging experience, which is to just use `stdout`. We also don't place restrictions on where the log file gets written to. We have a `/data` directory that's writable by the Kibana process, but this doesn't feel like the right place for log files.

Reply:
> I thought we were going to enable at most one of the audit loggers, based on config settings.

I'll be a little more direct: I believe this is the approach we need to take, defaulting to the legacy logger for the remainder of 7.x.

Reply:
> I thought we were going to enable at most one of the audit loggers, based on config settings.

I hadn't considered that, but it makes sense to me!

In that case, if we are logging to the console, do we want to avoid using the `debug` level? Otherwise our new audit logs will only show up if the user has `logging.verbose: true` or otherwise enables debug logs with the new logger config.

Our legacy audit logs use the `info` log level, so I'd assume we want to use that for our new audit logs too by default.

Reply:
> Our legacy audit logs use the info log level, so I'd assume we want to use that for our new audit logs too by default.

I agree, `info` makes sense to me as well.

Reply:
Good point, will change